### ISSUE TITLE
Class-level type annotations

### ISSUE THREAD

**Issue by IggShaman** [pytorch/pytorch#30695]
## 🐛 Bug

torch.jit.script(..) fails on a module with class-level type annotation `Optional[some-python-module]`

## To Reproduce

1. Initial piece of code:
```
from typing import Optional
import torch
from torch import nn
    
class Model(nn.Module):
    op: Optional[nn.ReLU]
    
    def __init__(self, op=None):
        super().__init__()
        self.op = op
    
    def forward(self, input):
        if self.op is not None:
            input = self.op(input)
        return input
```
2. This works:
```
with_op = Model(nn.ReLU())
with_op_js = torch.jit.script(with_op)
```
3. This fails:
```
noop = Model()
noop_js = torch.jit.script(noop)
```

Other annotations such as `Optional[nn.Module]`, `Optional[custom-python-class]` don't work either.

Traceback:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-f0219e553c87> in <module>
     19 
     20 noop = Model()
---> 21 noop_js = torch.jit.script(noop)
     22 
     23 #with_op = Model(nn.ReLU())

~/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/__init__.py in script(obj, optimize, _frames_up, _rcb)
   1201 
   1202     if isinstance(obj, torch.nn.Module):
-> 1203         return torch.jit.torch.jit._recursive.recursive_script(obj)
   1204 
   1205     qualified_name = _qualified_name(obj)

~/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/_recursive.py in recursive_script(mod, exclude_methods)
    171     filtered_methods = filter(ignore_overloaded, methods)
    172     stubs = list(map(make_stub, filtered_methods))
--> 173     return copy_to_script_module(mod, overload_stubs + stubs)
    174 
    175 

~/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/_recursive.py in copy_to_script_module(original, stubs)
     68         item = getattr(original, name)
     69         if name in class_annotations:
---> 70             the_type = torch.jit.annotations.ann_to_type(class_annotations[name])
     71         else:
     72             the_type = torch._C._jit_try_infer_type(item)

~/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/annotations.py in ann_to_type(ann, resolver)
    237     elif is_optional(ann):
    238         if issubclass(ann.__args__[1], type(None)):
--> 239             return OptionalType(ann_to_type(ann.__args__[0]))
    240         else:
    241             return OptionalType(ann_to_type(ann.__args__[1]))

~/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/annotations.py in ann_to_type(ann, resolver)
    256         if the_type is not None:
    257             return the_type
--> 258     raise ValueError("Unknown type annotation: '{}'".format(ann))
    259 
    260 

ValueError: Unknown type annotation: '<class 'torch.nn.modules.activation.ReLU'>'
```

## Expected behavior

Expected to obtain a version of a module with 'op' disabled.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 19.10
GCC version: (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2080 SUPER
Nvidia driver version: 440.36
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] pytorch                   1.3.1           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.4.2                py37_cu101    pytorch
[conda] torchviz                  0.0.1                    pypi_0    pypi
```

cc @suo

**Comment by eellison** [comment]
cc oncall when posted @wanchaol

**Comment by wanchaol** [comment]
Hey It's not about `Optional[Module]` not works as type annotation, its mainly because `Module` is not a valid type annotation. TorchScript type hints right now is only a subset of Python 3 typehints, class-level type annotation is not supported for both of these, so we could not annotate it like this way, submodules does not need to be annotated in this case.

Mypy style type check has a wider range of type support than Python 3 Typehints and TorchScript type hints. But this is something that we are not targeting on.

**Comment by IggShaman** [comment]
My use case here is to be able to jit.script-ify a custom Module which can have optional sub-module(s). This may be useful when for example a custom module class (with activation as the last op) is used to build a chain of modules, with the last instance not doing activation.

Presently, in order to achieve something like this and in order for JIT to work, I have to either use a proxy, "pass-through" module (for when activation is used), or a "noop" module (when activation is disabled).

Here's a brief, working example:

```
import torch
from torch import nn

class OpModule(nn.Module):
    """Passes inputs through saved "op"."""

    def __init__(self, op):
        super().__init__()
        self.op = op
        self.is_op = True

    def forward(self, input):
        return self.op(input)

class Noop(nn.Module):
    """Noop: doesn't do anything with its inputs. Has a flag to denote that this is not a real op."""
    def __init__(self):
        super().__init__()
        self.is_op = False

    # NOTE: this method needs to have exact same signature as OpModule.forward.
    # jit.scrit(..) will fail otherwise.
    def forward(self, input):
        return input

class WorkingBlock(nn.Module):

    def __init__(self, main_block, activation=None):
        super().__init__()

        self.main_block = main_block
        if activation is None:
            self.activation = Noop()
        else:
            self.activation = OpModule(activation)

    def forward(self, input):
        out = self.main_block(input)
        if self.activation.is_op:
            out = self.activation(out)
        return out


class FaultyBlock(nn.Module):

    def __init__(self, main_block, activation=None):
        super().__init__()

        self.main_block = main_block
        self.activation = activation

    def forward(self, input):
        out = self.main_block(input)
        if self.activation is not None:
            out = self.activation(out)
        return out

def make_test_network(block_class):
    return nn.Sequential(
        block_class(nn.Conv2d(3, 10, 3), nn.ReLU()),
        block_class(nn.Conv2d(10, 10, 3), nn.ReLU()),
        block_class(nn.Conv2d(10, 10, 3), activation=None),
    )

if __name__ == '__main__':
    net = make_test_network(WorkingBlock)
    net_scripted = torch.jit.script(net)
    result = net_scripted(torch.randn(16, 3, 100, 100))
    print(result.shape)

    net = make_test_network(FaultyBlock)
    net_scripted = torch.jit.script(net)
    result = net_scripted(torch.randn(16, 3, 100, 100))
    print(result.shape)
```

Presently, `FaultyBlock(something, activation=None)` will not register a sub-module, and a JIT-ed version doesn't even contain "activation" as object's attribute.

Is there a simpler workaround for this problem?

**Comment by wanchaol** [comment]
@IggShaman This does not has to be that complicated I think. You can just use None   together with modules. For example your original example can easily work with the following code (remove the type annotation and use None):
```python
from typing import Optional
import torch
from torch import nn
    
class Model(nn.Module):
    
    def __init__(self, op=None):
        super().__init__()
        self.op = op
    
    def forward(self, input):
        if self.op is not None:
            input = self.op(input)
        return input

with_op = Model(nn.ReLU())
with_op_js = torch.jit.script(with_op)

with_none = Model(None)
with_none_js = torch.jit.script(with_none)
print(with_none_js.graph)
```

**Comment by IggShaman** [comment]
@wanchaol your code sample gives the following error -

```
Traceback (most recent call last):
  File "g.py", line 20, in <module>
    with_none_js = torch.jit.script(with_none)
  File "/home/igor/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/__init__.py", line 1203, in script
    return torch.jit.torch.jit._recursive.recursive_script(obj)
  File "/home/igor/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/_recursive.py", line 173, in recursive_script
    return copy_to_script_module(mod, overload_stubs + stubs)
  File "/home/igor/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/_recursive.py", line 95, in copy_to_script_module
    torch.jit._create_methods_from_stubs(script_module, stubs)
  File "/home/igor/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/jit/__init__.py", line 1423, in _create_methods_from_stubs
    self._c._create_methods(self, defs, rcbs, defaults)
RuntimeError: 
module has no attribute 'op':
at g.py:12:11
    def forward(self, input):
        if self.op is not None:
           ~~~~~~~ <--- HERE
            input = self.op(input)
        return input
```

**Comment by wanchaol** [comment]
@IggShaman which version are you in? Can you try the nightly build or latest master? On my side it works fine:

```
from typing import Optional
import torch
from torch import nn
    
class Model(nn.Module):
    
    def __init__(self, op=None):
        super().__init__()
        self.op = op
    
    def forward(self, input):
        if self.op is not None:
            input = self.op(input)
        return input

with_op = Model(nn.ReLU())
with_op_js = torch.jit.script(with_op)
print(with_op_js.graph)

with_none = Model(None)
with_none_js = torch.jit.script(with_none)
print(with_none_js.graph)
```

output:

```graph(%self : __torch__.___torch_mangle_1.Model,
      %input.1 : Tensor):
  %5 : __torch__.torch.nn.modules.activation.ReLU = prim::GetAttr[name="op"](%self)
  %input.3 : Tensor = prim::CallMethod[name="forward"](%5, %input.1) # <ipython-input-2-5ddb29ed2838>:13:20
  return (%input.3)

graph(%self : __torch__.___torch_mangle_2.Model,
      %input.1 : Tensor):
  return (%input.1)
```

**Comment by IggShaman** [comment]
@wanchaol looks like it is fixed now in 1.4-dev. I came across this issue with 1.3.1 a few weeks back. Thanks for the help!
